lua/ttt2/extensions/math.lua

```diff
@@ -19,3 +19,48 @@ function math.ExponentialDecay(halflife, dt)
     -- ln(0.5) = -0.69..
     return exp((-0.69314718 / halflife) * dt)
 end
+
+---
+-- Gets the index of an item in the provided table, weighted according to the weights (derived from get_weight).
+-- @param table tbl The array of items to find a weighted item in.
+-- @param function get_weight Called as get_weight(item, index). Must return number.
+-- @return number
+-- @realm shared
+function math.random_weighted(tbl, get_weight)
+    -- There are several possible ways to get a weighted item. The most obvious is to simply include an item in
+    -- the table N times, where N is an integer proportional to the weight. This, however, requires maintaining
+    -- that table, which may be undesirable.
+    --
+    -- For a more friendly API, we instead compute the sum, generate a random number from 0 to that sum, then
+    -- take the first item whose weight prefix-sum is greater than that random number. This requires 2 passes
+    -- over the table, but enables processing on a normal table, with arbitrary weight storage.
+
+    -- Special case short arrays, because they may otherwise cause problems
+    if #tbl == 0 then
+        return nil
+    end
+    if #tbl == 1 then
+        return 1
+    end
+
+    -- first, compute the sum weight
+    local sum = 0
+    for k, v in pairs(tbl) do
+        sum = sum + get_weight(v, k)
+    end
+
+    -- get the random number
+    local rand = math.random(0, sum - 1)
+
+    -- now do the prefix-sum for the final value
+    sum = 0
+    for k, v in pairs(tbl) do
+        sum = sum + get_weight(v, k)
```

Comment:
Optional: Just as a sidenote, you could also do this computation while summing up all weights and do get some precomputed prefix sum table. 

I don't think this would make any real difference in performance in our use cases, but i just wanted to let you know!

Bonus effect would be that the weights are ensured to not have changed in between those `get_weight` invocations :)

Reply:
That would then cause a potentially large extra table allocation for large input tables. I don't think it would necessarily have any meaningful performance improvement for any input size. Would need to measure to be sure, which can be done if there appears to be perf issues in role distribution.